### PR TITLE
feat: add model-aware learned adversarial attacks

### LINKED ISSUE #164: [Feature]: Model-aware learned adversarial distortion generator (gradient-based attacks)
## Problem

The learned adversarial path (`nightmarenet/distortions/learned.py`) uses a generic MLM (DistilBERT) to generate perturbations via attention-weighted token replacement. It's not model-aware - it targets tokens that the MLM considers important, not tokens that the TARGET model being trained is most sensitive to. This means the adversarial examples don't exploit the specific weaknesses of the model under training, limiting the effectiveness of the nightmare phase.

Currently, `LearnedAdversarialGenerator`:
- Uses `_get_token_importance()` based on MLM attention weights (generic, not target-model-specific)
- Replaces important tokens with MLM predictions (plausible but not maximally adversarial for the target)
- Is disabled by default (`learned: 0.0` in config) partly because it's slow (loads a full MLM per call) and not demonstrably better than rule-based distortions

**Relevant files:**
- `nightmarenet/distortions/learned.py` (`LearnedAdversarialGenerator` class - current implementation)
- `nightmarenet/distortions/nightmare.py` (calls learned distortion when `learned > 0.0`)
- `nightmarenet/distortions/registry.py` (plugin registration - new strategies register here)
- `configs/default.yaml` (`distortion.adversarial.learned: 0.0`, `learned_model: "distilbert-base-uncased"`)

## Proposed Solution

Make the learned adversarial generator model-aware by using the TARGET model's gradients (not a separate MLM's attention) to identify vulnerable tokens:

1. **Gradient-based importance scoring** - Replace the current attention-based `_get_token_importance()` with gradient-based importance: compute `grad(loss, embeddings)` on the target model, use the L2 norm of per-token embedding gradients as importance scores. Tokens where small changes in embeddings cause large loss changes are the most vulnerable.

2. **Gradient-guided substitution** - For the top-k important tokens, use the embedding gradient direction to find substitutions that INCREASE loss (not just plausible MLM replacements). Project gradient onto the embedding matrix to find the nearest token in the adversarial direction.

3. **Caching** - Cache generated adversarial examples per cycle. As the model improves across cycles, previously adversarial examples may no longer be effective, so regenerate each cycle but keep the cache for curriculum scheduling (replay earlier adversarial examples alongside new ones).

4. **Config extension:**
   ```yaml
   distortion:
     adversarial:
       learned: 0.3
       learned_model: "distilbert-base-uncased"  # fallback MLM when target model unavailable
       learned_strategy: "gradient"  # "attention" (current) | "gradient" (new, model-aware)
       learned_cache: true            # cache adversarial examples across cycles
   ```

## Acceptance Criteria

- [ ] `LearnedAdversarialGenerator` accepts an optional `target_model` parameter for gradient-based importance
- [ ] Gradient-based importance: `grad(loss, embeddings).norm(dim=-1)` used to rank tokens
- [ ] Gradient-guided substitution: project gradient onto embedding matrix for adversarial token selection
- [ ] Falls back to current attention-based method when `target_model` not provided (backward compatible)
- [ ] Config: `learned_strategy: "gradient"` selects new method, `"attention"` preserves current behavior
- [ ] Adversarial example caching across cycles (configurable)
- [ ] Benchmark: compare gradient-based vs attention-based on robustness improvement for same compute budget
- [ ] Test: gradient-based produces different (ideally more adversarial) outputs than attention-based for same input
- [ ] Test: deterministic under fixed seed
- [ ] No regression when `learned: 0.0` (disabled path unchanged)
- [ ] Performance: gradient-based should not be >3x slower than attention-based per sample

## Technical Notes

- **Gradient computation requires `torch.enable_grad()` context** since the nightmare phase may run under `torch.no_grad()` for evaluation. Ensure the gradient computation is isolated.
- **Memory:** Computing gradients on the full model for each sample is expensive. Use `torch.autograd.grad` with `create_graph=False` to avoid storing the computation graph.
- **The target model is available in `NightmarePhase`** - it's `self.model`. The challenge is passing it through to the distortion generator, which currently operates at the data-generation level (before training). Consider: generate adversarial examples at the START of each nightmare epoch using the current model weights, not during data preparation.

## Scope

Large. ~6-8 hours. Requires understanding of PyTorch autograd, embedding spaces, and the existing distortion pipeline. Research-level implementation.

---

> **Important:** Please read [CONTRIBUTING.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/CONTRIBUTING.md) in full before requesting assignment. Assignment requests without a technical approach will not be considered. PRs demonstrating thorough research, benchmarks, or creative approaches beyond minimum criteria may earn bonus XP labels.

### PR DESCRIPTION
**Pull request by Jidnyasa-P**
## Summary

Adds a target-model-aware learned adversarial distortion strategy using embedding gradients and gradient-guided token substitution.

## Motivation

The previous learned distortion strategy used attention scores from a separate masked language model. It therefore selected tokens that were important to the fallback MLM rather than tokens to which the model currently being trained was most sensitive.

This change introduces a model-aware gradient strategy while retaining the existing attention strategy as a backward-compatible fallback.

Closes #164

## Changes

* Extended `LearnedAdversarialGenerator` with optional target-model and target-tokenizer support.
* Added token-importance scoring using the L2 norm of the loss gradient with respect to input embeddings.
* Added gradient-guided substitution using projection onto the target model’s embedding matrix.
* Added support for sequence-classification, masked-LM, and causal-LM target outputs.
* Added automatic fallback to the existing attention-based strategy when a compatible target model is unavailable.
* Added deterministic generation using a fixed seed.
* Added configurable per-cycle adversarial-example caching.
* Preserved cache entries when current model weights or model instances change across cycles.
* Wired the target model and tokenizer into nightmare dataset generation.
* Regenerated gradient-based nightmare examples at the beginning of later training cycles.
* Preserved the original preparation order when learned gradient distortion is disabled.
* Updated the standalone `nightmarenet-train` workflow with the same model-aware behavior.
* Added configuration options:

  * `learned_strategy`
  * `learned_cache`
  * `learned_seed`
* Added focused unit tests and an attack-performance benchmark script.
* Added research documentation covering the algorithm, configuration, cycle integration, and benchmark methodology.

## Acceptance Criteria

* [x] `LearnedAdversarialGenerator` accepts an optional target model.
* [x] Token importance uses `grad(loss, embeddings).norm(dim=-1)`.
* [x] Replacement selection projects gradients onto the embedding matrix.
* [x] Missing target models fall back to the attention strategy.
* [x] `learned_strategy: "gradient"` and `"attention"` are supported.
* [x] Adversarial examples are cached by training cycle.
* [x] Gradient and attention strategies produce different outputs in tests.
* [x] Generation is deterministic with a fixed seed.
* [x] `learned: 0.0` leaves the learned path disabled.
* [ ] Full robustness comparison completed using an identical compute budget.
* [ ] Measured gradient generation is no more than three times slower than attention generation.

## Testing

```bash
pytest tests/test_learned_adversarial.py -v
pytest tests/test_distortions.py -v
ruff check nightmarenet/ scripts/ tests/
mypy nightmarenet/ --ignore-missing-imports
pytest tests/
```

Focused implementation results:

* 8 model-aware learned-adversarial tests passed.
* 40 existing distortion regression tests passed.
* Ruff passed on all modified Python files.

## Benchmark

Command:

```bash
python scripts/benchmark_learned_adversarial.py \
  --model distilbert-base-uncased \
  --strength 0.7 \
  --repeats 3 \
  --output artifacts/learned-adversarial-benchmark.json \
  --enforce-slowdown
```

Measured results:

{
  "model": "distilbert-base-uncased",
  "device": "cpu",
  "strength": 0.7,
  "seed": 42,
  "repeats": 3,
  "attention": {
    "strategy": "attention",
    "samples": 24,
    "mean_seconds_per_sample": 0.09725531250008619,
    "median_seconds_per_sample": 0.09746235000056913,
    "mean_target_loss_before": 2.7623347640037537,
    "mean_target_loss_after": 3.080944538116455,
    "mean_target_loss_increase": 0.3186097741127014,
    "changed_outputs": 24
  },
  "gradient": {
    "strategy": "gradient",
    "samples": 24,
    "mean_seconds_per_sample": 0.09642325000064982,
    "median_seconds_per_sample": 0.08019420000164246,
    "mean_target_loss_before": 2.7623347640037537,
    "mean_target_loss_after": 3.7494084537029266,
    "mean_target_loss_increase": 0.987073689699173,
    "changed_outputs": 24
  },
  "gradient_to_attention_slowdown": 0.9914445547698422,
  "within_requested_3x_budget": true
}

Screenshot 2026-07-16 213210
Screenshot 2026-07-16 215716
Screenshot 2026-07-16 215847

## Documentation

* Added `docs/research/model-aware-learned-adversarial.md`.
* Updated configuration comments for the learned strategies.
* Added benchmark and reproducibility instructions.

## AI Disclosure

This PR includes AI-assisted code generation. I reviewed every changed line, ran the documented tests, and can explain the gradient computation, embedding projection, caching, and training-cycle integration.

## Summary by CodeRabbit

* **New Features**
  * Added model-aware “learned adversarial” generation with selectable **attention** vs **gradient** strategies.
  * Expanded learned adversarial configuration with strategy selection, cache controls, deterministic seeding, and cycle-aware behavior.
  * Training now integrates the target model/tokenizer and updates learned adversarial generation per cycle.
  * Added a benchmark script to compare attention vs gradient performance and effectiveness.
* **Documentation**
  * Added a guide describing strategies, configuration options, verification steps, and robustness experiment guidance.
* **Bug Fixes**
  * Improved resilience by gracefully falling back to attention when gradient generation is unavailable or fails.
* **Tests**
  * Added offline deterministic tests covering strategy outputs, determinism, cache partitioning, fallback behavior, and gradient-safety.

### CODE REVIEW

**Review by Adit-Jain-srm — CHANGES_REQUESTED** [review]
Resolve code-rabbit comments and CI / lint checks errors.
Through implementation, great work @Jidnyasa-P

**Comment by Jidnyasa-P** [comment]
> Resolve code-rabbit comments and CI / lint checks errors. Through implementation, great work @Jidnyasa-P

Done @Adit-Jain-srm

**Review by Adit-Jain-srm — APPROVED** [review]
LGTM!
